pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -421,9 +421,48 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 		}
 	}
 
+	as, err := artifacts.InitializeArtifactStorage(c.Images, pr, pipelineSpec, c.KubeClientSet, c.Logger)
+	if err != nil {
+		c.Logger.Infof("PipelineRun failed to initialize artifact storage %s", pr.Name)
+		return err
+	}
+
+	// When the pipeline run is stopping, we don't schedule any new task and only
+	// wait for all running tasks to complete and report their status
+	if !pipelineState.IsStopping() {
```

Comment:
The diff does not help very much, but the only changes here are:
- `if !pipelineState.IsStopping()`
- the code that used to be here moved "as-is" into `runNextSchedulableTask`

Reply:
May be this could be simplified and keep it as is, instead change `GetNextTasks()` such that no tasks are returned if pipeline is in stopping mode:

```
func (state PipelineRunState) GetNextTasks(candidateTasks map[string]struct{}) []*ResolvedPipelineRunTask {
	tasks := []*ResolvedPipelineRunTask{}
	if state.IsStopping() {
		return tasks
	}
```

With an empty execution queue, everything from `runNextSchedulableTask` is escaped except `InitializeArtificatStorage` (could be surrounded by the condition)

Dont worry if you are not comfortable changing it, I will have to change it in `finally` PR, as final tasks have to be executed after pipeline is stopping and once all running tasks are done.